src/hotspot/share/nmt/memLogRecorder.cpp

```diff
@@ -0,0 +1,855 @@
+#include "utilities/nativeCallStack.hpp"
+#include "utilities/permitForbiddenFunctions.hpp"
+#include "utilities/vmError.hpp"
+
+bool NMTRecorder_Locker::_safe_to_use = false;
+
+#if defined(LINUX) || defined(__APPLE__)
+
+#include <locale.h>
+#include <string.h>
+#include <errno.h>
+
+#if defined(LINUX)
+#include <malloc.h>
+#elif defined(__APPLE__)
+#include <malloc/malloc.h>
+#endif
+
+#include <pthread.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#define LD_FORMAT "%'ld"
+
+static void* raw_realloc(void* old, size_t s) { return permit_forbidden_function::realloc(old, s); }
+#if defined(LINUX)
+static size_t raw_malloc_size(void* ptr)      { return ::malloc_usable_size(ptr); }
+#elif defined(_WIN64)
+static size_t raw_malloc_size(void* ptr)      { return ::_msize(ptr); }
+#elif defined(__APPLE__)
+static size_t raw_malloc_size(void* ptr)      { return ::malloc_size(ptr); }
+#endif
+
+
+#define NMT_HEADER_SIZE 16
+
+NMT_MemoryLogRecorder NMT_MemoryLogRecorder::_recorder;
+NMT_VirtualMemoryLogRecorder NMT_VirtualMemoryLogRecorder::_recorder;
+
+void NMT_LogRecorder::initialize() {
+  char* NMTRecordMemoryAllocations = getenv("NMTRecordMemoryAllocations");
+  if (NMTRecordMemoryAllocations != nullptr) {
+    long count = atol(NMTRecordMemoryAllocations);
+    if (count == 0) {
+      count = strtol(NMTRecordMemoryAllocations, nullptr, 16);
+    }
+    NMT_MemoryLogRecorder::initialize(count);
+  }
+  char* NMTRecordVirtualMemoryAllocations = getenv("NMTRecordVirtualMemoryAllocations");
+  if (NMTRecordVirtualMemoryAllocations != nullptr) {
+    long count = atol(NMTRecordVirtualMemoryAllocations);
+    if (count == 0) {
+      count = strtol(NMTRecordVirtualMemoryAllocations, nullptr, 16);
+    }
+    NMT_VirtualMemoryLogRecorder::initialize(count);
+  }
+}
+
+void NMT_LogRecorder::finish() {
+  if (!NMT_MemoryLogRecorder::instance()->done()) {
+    NMT_MemoryLogRecorder::instance()->finish();
+  }
+  if (!NMT_VirtualMemoryLogRecorder::instance()->done()) {
+    NMT_VirtualMemoryLogRecorder::instance()->finish();
+  }
+}
+
+void NMT_LogRecorder::replay() {
+  char* NMTBenchmarkRecordedPID = getenv("NMTBenchmarkRecordedPID");
+  if (NMTBenchmarkRecordedPID != nullptr) {
+    int pid = atoi(NMTBenchmarkRecordedPID);
+    NMT_MemoryLogRecorder::instance()->replay(pid);
+    NMT_VirtualMemoryLogRecorder::instance()->replay(pid);
+    os::exit(0);
+  }
+}
+
+void NMT_LogRecorder::init() {
+  _threads_names_size = 1;
+  _threads_names = (thread_name_info*)raw_realloc(_threads_names, _threads_names_size*sizeof(thread_name_info));
+  _done = true;
+  _count = 0;
+}
+
+void NMT_LogRecorder::get_thread_name(char* buf) {
+#if defined(__APPLE__)
+  if (pthread_main_np()) {
+    strcpy(buf, "main");
+  } else {
+    pthread_getname_np(pthread_self(), buf, MAXTHREADNAMESIZE);
+  }
+#elif defined(LINUX)
+  pthread_getname_np(pthread_self(), buf, MAXTHREADNAMESIZE);
+#elif defined(_WIN64)
+  // TODO: NMT_LogRecorder::thread_name
+#endif
+}
```

Comment:
`Thread::current()->name()`

Reply:
`Thread::current()->name()` has different semantics. It uses names for partially constructed threads, like `<no-name - thread is attaching>`, and `Unknown thread`, and since we save the thread name the 1st time, we currently get a name and we will be stuck with it forever.

On the other hand, when `pthread_getname_np()` returns a string, we get a real final value.

To go with `Thread::current()->name()` I need to check for those special temp names and skip those, but I agree, it is worth to use platform independent API here.

I will make a change.